cyral/resource_cyral_integration_idp_saml.go

```diff
@@ -0,0 +1,165 @@
+package cyral
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+
+	"github.com/cyralinc/terraform-provider-cyral/client"
+)
+
+type CreateGenericSAMLRequest struct {
+	SAMLDraftId string                  `json:"samlDraftId"`
+	IdpMetadata *GenericSAMLIdpMetadata `json:"idpMetadata,omitempty"`
+
+	// Currently unused fields
+	IdpDescriptor *GenericSAMLIdpDescriptor `json:"idpDescriptor,omitempty"`
+}
+
+func (req *CreateGenericSAMLRequest) ReadFromSchema(d *schema.ResourceData) error {
+	req.SAMLDraftId = d.Get("saml_draft_id").(string)
+	if url := d.Get("idp_metadata_url").(string); url != "" {
+		req.IdpMetadata = &GenericSAMLIdpMetadata{
+			URL: url,
+		}
+	} else if xml := d.Get("idp_metadata_document").(string); xml != "" {
+		req.IdpMetadata = &GenericSAMLIdpMetadata{
+			XML: xml,
+		}
+	}
+	return nil
+}
+
+type CreateGenericSAMLResponse struct {
+	Integration GenericSAMLIntegration `json:"integration"`
+}
+
+func (resp *CreateGenericSAMLResponse) WriteToSchema(d *schema.ResourceData) error {
+	return resp.Integration.WriteToSchema(d)
+}
+
+func (resp *CreateGenericSAMLResponse) ReadFromSchema(d *schema.ResourceData) error {
+	return resp.Integration.ReadFromSchema(d)
+}
+
+type ReadGenericSAMLResponse struct {
```

Comment:
The reason I chose to make them different is because the response schemas are not the same. `Update` returns `{"id":"...", "identityProvider": {...}}`, whereas `Read` returns just `{"identityProvider": {...}}`.

Having this separation will also help adapt the provider in case there is an API change in the future that is asymmetric between `Read` and `Update`.

Reply:
The `Update` returns an [empty response](https://github.com/cyralinc/integrations/blob/main/proto/generic-saml/generic_saml.proto#L473), the ID of the request payload is just a path variable that is passed in the URL, so the body of the Update request and the Read response end up being the same. I mean, I'm also fine with using two different structures here, for the reasons you mentioned.

Reply:
Oops sorry, I got confused there for a moment. I will remove the `ID` field in UpdateGenericSAMLRequest. Thanks! I will maintain the distinction between read and updates, though, as I really believe it makes the code easier to read.